sdk/communication/azure-communication-chat/azure/communication/chat/_shared/user_credential.py

```diff
@@ -3,84 +3,121 @@
-
-    def __init__(self,
-                 token,  # type: str
-                 **kwargs
-                 ):
-        token_refresher = kwargs.pop('token_refresher', None)
-        communication_token_refresh_options = CommunicationTokenRefreshOptions(token=token,
-                                                                               token_refresher=token_refresher)
-        self._token = communication_token_refresh_options.get_token()
-        self._token_refresher = communication_token_refresh_options.get_token_refresher()
+    _DEFAULT_AUTOREFRESH_INTERVAL_MINUTES = 10
+
+    def __init__(self, token: str, **kwargs: Any):
+        if not isinstance(token, six.string_types):
+            raise TypeError("Token must be a string.")
+        self._token = create_access_token(token)
+        self._token_refresher = kwargs.pop('token_refresher', None)
+        self._refresh_proactively = kwargs.pop('refresh_proactively', False)
+        self._timer = None
         self._lock = Condition(Lock())
         self._some_thread_refreshing = False
+        if self._refresh_proactively:
+            self._schedule_refresh()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        if self._timer is not None:
+            self._timer.cancel()
 
     def get_token(self, *scopes, **kwargs):  # pylint: disable=unused-argument
         # type (*str, **Any) -> AccessToken
         """The value of the configured token.
         :rtype: ~azure.core.credentials.AccessToken
         """
 
-        if not self._token_refresher or not self._token_expiring():
+        if not self._token_refresher or not self._is_token_expiring_soon(self._token):
             return self._token
 
-        should_this_thread_refresh = False
+        self._update_token_and_reschedule()
+        return self._token
 
+    def _update_token_and_reschedule(self):
+        should_this_thread_refresh = False
         with self._lock:
-            while self._token_expiring():
+            while self._is_token_expiring_soon(self._token):
                 if self._some_thread_refreshing:
-                    if self._is_currenttoken_valid():
+                    if self._is_token_valid(self._token):
                         return self._token
-
-                    self._wait_till_inprogress_thread_finish_refreshing()
+                    self._wait_till_lock_owner_finishes_refreshing()
                 else:
                     should_this_thread_refresh = True
                     self._some_thread_refreshing = True
                     break
 
         if should_this_thread_refresh:
             try:
-                newtoken = self._token_refresher()  # pylint:disable=not-callable
-
+                new_token = self._token_refresher()
+                if not self._is_token_valid(new_token):
+                    raise ValueError(
+                        "The token returned from the token_refresher is expired.")
```

Comment:
Should we attempt to validate this in the client?
It seems we could just send the token at let the service reject it with some kind of auth error?

Reply:
We discussed it with Dom some time ago and agreed we should set clear expectations about how the callback method is supposed to by implemented by Contoso. It doesn't make any sense to allow expired tokens, and throwing an exception is, I think, the right thing to do in this case. Also, this way the code follows the "throw early, catch late" principle. Is it ok from your PoV?

Reply:
We don't think clock skew could be a problem where we raise on a token that might actually have been accepted?
Likewise in reverse - this wont prevent expired token being sent to the service, it will just filter out some of the _really_ expired tokens - so users will have to handle _both_ ValueErrors and Authentication errors in the scenario of expired tokens.

Reply:
This is a valid point and yes, it can happen. 
I wonder - Would you feel different about an argument null check? To me, they feel very similar - they're just a set of pre-conditions that help eliminate possible invalid scenarios, but never 100% guarantee that something won't fail later on.
The necessity of handling this type of errors still feels like a small price to pay for receiving a clear and concise message about the source of the error early on. 
I'm really curious about your PoV.

Reply:
There are similarities, whereas an argument check feels like it would be more of an error that would occur during development, causing someone to correct their code.
My feeling is that this could more likely occur due (depending on how the users callback acquires a token) to awkward timing, maybe a flaky network for example, that could result in the same code sometimes giving a ValueError and sometimes given a service AuthenticationError.
I guess another factor to consider here would be retry - if a ValueError is raised because the callback ran slowly for some reason - should that be retryable? Is that up to the user to retry? If instead it's rejected by the service - will that/should that be retried in the pipeline or by customers? (I don't think the pipeline automatically retries auth errors).

Reply:
hmm good question - if I send an expired token to the service, what status code would I receive?
It's possible this wont be retried, and simply raised directly to the customer.
We can double check the code path/pipeline retry policy to confirm :)

Reply:
@AikoBB informed me that you've discussed this one before she left for a vacation. -> resolving
![MicrosoftTeams-image](https://user-images.githubusercontent.com/9810625/166208368-9a16c997-405f-4e00-b68d-626ee28ab5e4.png)